### PR TITLE
add organization webhook support

### PR DESCRIPTION
**Pull request by a1012112796**
it's same with gh, ref:
https://docs.github.com/en/github-ae@latest/developers/webhooks-and-events/webhook-events-and-payloads#organization
https://docs.github.com/en/github-ae@latest/developers/webhooks-and-events/webhook-events-and-payloads#team
https://docs.github.com/en/github-ae@latest/developers/webhooks-and-events/webhook-events-and-payloads#membership

### CODE REVIEW

**Comment by zeripath** [comment]
please fix conflicts.

File: models/webhook.go [hunk 1]
```diff
@@ -629,10 +665,38 @@ func (h HookEventType) Event() string {
...
 	}
 	return ""
 }
 
+// HookEventLevel leve of a hook event
+type HookEventLevel int64
+
+const (
+	// HookEventLevelRepo all hook types can be used
+	HookEventLevelRepo HookEventLevel = iota
```
Comment (adelowo) [inline models/webhook.go]:
```suggestion
  HookEventLevelRepo HookEventLevel = iota + 1
```

iota along makes it zero and it feels kind of weird seeing a valid state as 0

File: models/webhook.go [hunk 2]
```diff
@@ -765,6 +830,15 @@ func FindRepoUndeliveredHookTasks(repoID int64) ([]*HookTask, error) {
 	return tasks, nil
 }
 
+// FindOrgUndeliveredHookTasks represents find the undelivered hook tasks of one org
```
Comment (adelowo) [inline models/webhook.go]:
```suggestion
// FindOrgUndeliveredHookTasks finds the undelivered hook tasks of an organisation
```

File: services/webhook/webhook.go [hunk 3]
```diff
@@ -95,11 +95,16 @@ func getPayloadBranch(p api.Payloader) string {
 }
 
 // PrepareWebhook adds special webhook to task queue for given payload.
-func PrepareWebhook(w *models.Webhook, repo *models.Repository, event models.HookEventType, p api.Payloader) error {
-	if err := prepareWebhook(w, repo, event, p); err != nil {
+func PrepareWebhook(w *models.Webhook, repo *models.Repository, org *models.User, event models.HookEventType, p api.Payloader) error {
```
Comment (zeripath) [inline services/webhook/webhook.go]: It probably would be simpler to have:

```go
func PrepareWebhook(w *models.Webhook, repo *models.Repository, event models.HookEventType, p api.Payloader) error {
  return PrepareRepoOrgWebhook(w, repo, nil, event, p)
}

func PrepareOrgWebhook(w *models.Webhook, org *models.User, event models.HookEventType, p api.Payloader) error {
  return PrepareRepoOrgWebhook(w, nil, org, event, p)
}

func PrepareRepoOrgWebhook(w *models.Webhook, repo *models.Repository, org *models.User, event models.HookEventType, p api.Payloader) error {
 ...
}
```

Then you would need most of the changes in modules/notification/webhook/webhook.go and elsewhere.

File: models/migrations/v186.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+package migrations
+
+import (
+	"fmt"
+
+	"xorm.io/xorm"
+)
+
+func addOrgIDHookTaskColumn(x *xorm.Engine) error {
```
Comment (lunny) [inline models/migrations/v186.go]: You don't need that, since webhook table has already the column.

File: models/webhook.go [hunk 5]
```diff
@@ -649,6 +713,7 @@ type HookResponse struct {
 type HookTask struct {
 	ID              int64 `xorm:"pk autoincr"`
 	RepoID          int64 `xorm:"INDEX"`
+	OrgID           int64 `xorm:"INDEX NOT NULL DEFAULT 0"`
```
Comment (lunny) [inline models/webhook.go]: As above

**Review by lunny — DISMISSED** [review]
per my comments

File: models/migrations/v186.go (hunk 4 above)
Comment (a1012112796) [reply models/migrations/v186.go]: @lunny  But then how to find undelivered orgs task?
see ``services/webhook/deliver.go``

**Comment by lunny** [comment]
Please resolve the conflicts.

File: models/migrations/v188.go [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+func addOrgIDHookTaskColumn(x *xorm.Engine) error {
+	type HookTask struct {
+		OrgID int64 `xorm:"INDEX NOT NULL DEFAULT 0"`
+	}
+
+	if err := x.Sync2(new(HookTask)); err != nil {
+		return fmt.Errorf("Sync2: %v", err)
+	}
+	return nil
```
Comment (lunny) [inline models/migrations/v188.go]: But you should update the `org_id` from `webhook` table.

Comment (a1012112796) [reply models/migrations/v188.go]: Not needed, an orgnization event means an event which only related to an orgniztion (include `new team`, `new org member` ..), not means the events which related to any repos in a orgniztion, for them, just use repo id is enough and org_id should be set as zero, or ``FindOrgUndeliveredHookTasks()`` will get wrong result.

**Comment by a1012112796** [comment]
mark as breaking, because this pull will expand the meaning of ``orgnization hook``, if user has add an orgnization hook and choose ``Trigger On all Events``, **BE CAREFULLY !!!**

**Comment by wxiaoguang** [comment]
I think there it is:

image

**Pull request closed without merging**